### ISSUE TITLE
Possible Memory Leak in owlBuild* Functions

### ISSUE THREAD

**Issue by pseudo-randomness** [owl-project/owl#188]
I'm modifying an app that uses the SampleRenderer class. I added the following function that replaces the current model with a new one. It works very well, but GPU memory increases each time I add new models until it eventually runs out of memory. nvidia-smi shows memory increasing by 2 to 14 Mb depending on the complexity of the model.

  void SampleRenderer::setModel(Model *newModel)
  {
    model = newModel;
    createTextures();
    buildAccel();

    owlBuildPrograms(context);
    owlBuildPipeline(context);
    owlBuildSBT(context);
  }

**Comment by ingowald** [comment]
Sukatuh - Are you sure that the memory leaks are in these three functions? It's absolutely possible that they do have some leaks, but I'd almost suspect that there's other leaks as well when you create buffers for the models' mesh vertices, indices, textures, etc. The 'optix7sample' was never intended to be a real "complete" renderer, so it doesn't have any of the clean exit code that would free any of that model data - so if you create a new model and aren't careful about freeing whatever has already been created there's a good chance that stuff is leaking in there.

That said, as said above it _is_ absoltely possible that the leak is in one of these three functions. Could you do me a favor and just put an infinite loop around exactly these three functoins (ie, without loading a new model), and see if you still see the leak still persists?

**Comment by ingowald** [comment]
Also a general comment: if you're only changing _geometry_ and _shading_ related stuff - not the programs itself - then there is no need to rebuild the pipeline or programs - only the accels and SBT.

**Comment by Sukatuh** [comment]
The app crashes in owlBuildBST when I comment out the owlBuildPrograms and owlBuildPipeline function calls.

**Comment by ingowald** [comment]
You still have to call it *once*, of course. I meant you don't need to _re_build it once you switch to _another_ model. 
But back to the original problem: did you put an infinite loop around and verified that the memory leak really is in these functions?

**Comment by Sukatuh** [comment]
The constructor builds the initial programs and pipeline. I commented out the code only in the setModel function that is called when I want to change models.

Running with an infinite loop
Test 1
Comment out model=newModel, createTexture, and buildAccel: no change in memory usage
Test 2
Comment out model and createTexture (uses original model): +2 Mbs per call
Test 3
Comment out model and buildAccel: no change in memory usage

**Comment by ingowald** [comment]
So as expected, that points to the memory leak(s) being in that SampleRenderer's 'buildAccel()', not in the owlBuild* functions. And if you look at that function in the sample renderer, it does exactly what I suspected in my original answer ("I'd almost suspect that there's other leaks as well when you create buffers for the models' mesh vertices, indices, textures, etc.") - given a model (no matter whether it's a newly loaded one or not), this function creates all kind of owl buffers to upload vertices, indices, etc, build acceleration structures, etc, for that model .... and those get never freed. If you call it again - with either same or other model - it will simply re-allocate that stuff anew, never free it, and there's your leak.

As mentioned above, that sampler renderer was never intended to be a full renderer that can do things like changing models, etc; if you do want to use that as a template for your own you'll have to make sure that you free this memory. But it's not a leak in _OWL_, that's a leak in the app that uses owl.

**Comment by pseudo-randomness** [comment]
I modified the app to create a new SampleRenderer (SR) instance when I change models instead of replacing the model to an existing SR instance. I added the following destructor to the SampleRenderer class. 

  ```
SampleRenderer::~SampleRenderer()
  {
    if (context != nullptr)
    {
      owlContextDestroy(context);
      context = nullptr;
    }
  }
```
The owlContextDestroy function does a very good job of releasing GPU memory, but the class is still leaking ~150K bytes every time I change models. The main code in the constructor is

```
createTextures();
buildAccel();

owlBuildPrograms(context);
owlBuildPipeline(context);
owlBuildSBT(context);
```

When I comment out the last 3 lines, memory usage doesn't increase when I change models in an infinite loop; destroying the context frees GPU memory allocated by buildAccel. However, when I add owlBuildPrograms(context), the app leaks ~150K every time I create a new instance of SR (and free the old). owlBuildPrograms is allocating GPU memory that owlContextDestroy doesn't release.

**Comment by ingowald** [comment]
K; I'll have a look. I was (and am) fairly sure that there's no leakage in building or allocating geoms or accels, because we do that all the time in our own owl applications; but building _programs_ is something we'd usually not do more than once, so that's indeed something that's not well tested. As you said the context that that these programs were built in should release them when that context dies; if it doesn't then that's a bug and needs to get fixed. I'll be travelling the next few days but will look into that when I get back.

And just to explain what I meant when I implied that "we usually don't do that": In our own use cases we'd typiocally create a context only once, and build the programs only once (because next model will use the same programs, right?); then when the geometry changes you'd use functions like `owlGroupRelease`, `owlGeomRelase' and `owlTextureDestroy` to free existing geometry before loading/creating new one ... but within the same existing context, and without building new modules or programs. Killing the entire context and rebuilding programs feels a bit heavy-weight... BUT it should still work, and if it doesn't it needs to get fixed.

**Comment by ingowald** [comment]
confirmed there's a (modestly small) memory leak in owlbuildprograms. No idea where this is coming from (yet), but confirm i see a steady increase in gpu memory use if i just run that in a loop that creates context, builds progs, and destroys context.

**Comment by ingowald** [comment]
(in case you're wondering : flight got delayed ... )

It looks like this might be related to the "boundsModule", which is a CUDA module that get created for any potential bounds programs (optix doesn't have the concept of a bounds program, so i have to do that through cuda). This gets created, but apparently not destroyed; and since it's not part of the _optix_ context it also doesn't get destroyed when the optix context gets destroyed.
Still verifying that that's the culprit, but as of right now that looks to be the one.

**Comment by ingowald** [comment]
K; just pushed a presumed fix to devel branch. In my test case I took the optix7course sample renderer, and ran the constructor code (buildaccel, buildprograms, etc) in an infinite loop, every time creating context, running constructor body, and freeing .... and with that fix in devel I no longer see gpu memory increasing this way.

Could you test devel and let me know if that fixes the issue? If so i'll also merge into master.

**Comment by pseudo-randomness** [comment]
The change to Module.cpp fixed the memory leak. Thank you
